sys/linux/fs_ioctl_btrfs.txt

```diff
@@ -1,7 +1,426 @@
+ioctl$BTRFS_IOC_DEV_REPLACE(fd fd, cmd const[BTRFS_IOC_DEV_REPLACE], arg ptr[inout, btrfs_ioctl_dev_replace_args])
+
+#the arg should be of length 3. Not sure how to specify that
+ioctl$BTRFS_IOC_GET_SUPPORTED_FEATURES(fd fd, cmd const[BTRFS_IOC_GET_SUPPORTED_FEATURES], arg ptr[out, btrfs_ioctl_feature_flags])
+
+ioctl$BTRFS_IOC_GET_FEATURES(fd fd, cmd const[BTRFS_IOC_GET_FEATURES], arg ptr[out, btrfs_ioctl_feature_flags])
+ioctl$BTRFS_IOC_SET_FEATURES(fd fd, cmd const[BTRFS_IOC_SET_FEATURES], arg ptr[in, btrfs_ioctl_feature_flags])
+ioctl$BTRFS_IOC_GET_SUBVOL_INFO(fd fd, cmd const[BTRFS_IOC_GET_SUBVOL_INFO], arg ptr[out, btrfs_ioctl_get_subvol_info_args])
+ioctl$BTRFS_IOC_GET_SUBVOL_ROOTREF(fd fd, cmd const[BTRFS_IOC_GET_SUBVOL_ROOTREF], arg ptr[inout, btrfs_ioctl_get_subvol_rootref_args])
+ioctl$BTRFS_IOC_INO_LOOKUP_USER(fd fd, cmd const[BTRFS_IOC_INO_LOOKUP_USER], arg ptr[inout, btrfs_ioctl_ino_lookup_user_args])
+
+resource transid[int64]
+resource devid[int64]
+# This should be a resource but it's not returned by any syscall. What should we do about it?
+# resource qgroupid[int64]
+resource treeid[int64]
+resource dirid[int64]
+
+define BTRFS_PATH_MAX	BTRFS_PATH_NAME_MAX+1
+
+#This pads the struct to 1032 bytes. It was originally meant to pad to
+#1024 bytes, but when adding the flags field, the padding calculation
+#was not adjusted.
+define BTRFS_DEV_PAD	128-2-BTRFS_DEV_STAT_VALUES_MAX
+
+define BTRFS_VOL_MAX	BTRFS_VOL_NAME_MAX + 1
+
+#add a padding since fd in this struct should be 64 bits (is this the correct way?)
+btrfs_ioctl_vol_args {
+	fd	fd
+	dummy	const[0, int32]
+	name	array[int8, BTRFS_PATH_MAX]
+}
+
+btrfs_ioctl_vol_args_v2 {
+	fd	fd
+	transid	transid
+	flags	flags[btrfs_ioctl_vol_args_v2_flags, int64]
+	u1	btrfs_args_v2_u1
+	u2	btrfs_args_v2_u2
+}
+
+btrfs_args_v2_u1 [
+	inherit	btrfs_args_v2_u1_s1
+	unused	array[int64, 4]
+]
+
+btrfs_args_v2_u2 [
+	name		array[int8, BTRFS_PATH_MAX]
+	devid		devid
+	subvolid	int64
+]
+
+btrfs_args_v2_u1_s1 {
+	size		bytesize[qgroup_inherit, int64]
+	qgroup_inherit	ptr[in, btrfs_qgroup_inherit]
+}
+
+btrfs_qgroup_inherit {
+	flags		flags[btrfs_qgroup_inherit_flags, int64]
+	num_qgroups	int64
+	num_ref_copies	int64
+	num_excl_copies	int64
+	lim		btrfs_qgroup_limit
+}
+
+btrfs_qgroup_limit {
+	flags		flags[btrfs_qgroup_limit_flags, int64]
+	max_rfer	int64
+	max_excl	int64
+	rsv_rfer	int64
+	rsv_excl	int64
+}
+
+btrfs_ioctl_defrag_range_args {
+	start		int64
+	len		int64
+	flags		flags[btrfs_ioctl_defrag_range_args_flags, int64]
+	extent_thresh	int32
+	compress_type	flags[btrfs_compression_type, int32]
+	unused		array[int32, 4]
+}
+
+btrfs_ioctl_fs_info_args {
+	max_id		int64
+	num_devices	int64
+	fsid		array[int8, BTRFS_FSID_SIZE]
+	nodesize	int32
+	sectorize	int32
+	clone_alignment	int32
+	reserved32	int32
+	reserverxd	array[int64, 122]
+}
+
+btrfs_ioctl_dev_info_args {
+	devid		devid
+	uuid		array[int8, BTRFS_UUID_SIZE]
```

Comment:
I definitely agree that uuid should be a resource. However as far as I know resource can only be integer. What would be the best way to cope with this?

Reply:
Yes, currently only integers are supported as resources. How does kernel generate them? Maybe we can guess them? Or provoke kernel to use known values?
Since BTRFS_UUID_SIZE is 16, the best option is to represent it as 2 int64 (uuid_lo and uuid_hi), with some luck fuzzer engine can take a parallel pair and pass it as input. We will also need to plumb packed attribute somewhere to force 1 byte alignment.
Also add a `NEED:` comment explaining what support we are missing to express this properly.

Reply:
Unresolving, so we don't lose track of this improvement.